File: 22-Binary-Trees-Miscellaneous/cpp/01-binary-tree-to-doubly-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    TreeNode* prev = nullptr;
    TreeNode* head = nullptr;

    void inorder(TreeNode* node) {
        if (!node) return;

        inorder(node->left);

        if (prev == nullptr) {
            head = node;
        } else {
            prev->right = node;
            node->left = prev;
        }
        prev = node;

        inorder(node->right);
    }

public:
    TreeNode* bToDLL(TreeNode* root) {
        inorder(root);
        return head;
    }
};

File: 22-Binary-Trees-Miscellaneous/python/01_binary_tree_to_doubly_linked_list.py
from typing import Optional

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def bToDLL(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        self.prev = None
        self.head = None

        def inorder(node: Optional[TreeNode]):
            if not node:
                return

            inorder(node.left)

            if not self.prev:
                self.head = node
            else:
                self.prev.right = node
                node.left = self.prev
            self.prev = node

            inorder(node.right)

        inorder(root)
        return self.head
